src/Runner.Worker/Handlers/CompositeActionHandler.cs

```diff
@@ -268,13 +277,57 @@ private async Task RunStepsAsync(List<IStep> embeddedSteps)
                     step.ExecutionContext.Complete(TaskResult.Failed);
                 }
 
-                await RunStepAsync(step);
+                // Evaluate condition
+                step.ExecutionContext.Debug($"Evaluating condition for step: '{step.DisplayName}'");
+                var conditionTraceWriter = new ConditionTraceWriter(Trace, step.ExecutionContext);
+                var conditionResult = false;
+                var conditionEvaluateError = default(Exception);
+                if (HostContext.RunnerShutdownToken.IsCancellationRequested)
+                {
+                    step.ExecutionContext.Debug($"Skip evaluate condition on runner shutdown.");
+                }
+                else
+                {
+                    try
+                    {
+                        var templateEvaluator = step.ExecutionContext.ToPipelineTemplateEvaluator(conditionTraceWriter);
+                        var condition = new BasicExpressionToken(null, null, null, step.Condition);
+                        conditionResult = templateEvaluator.EvaluateStepIf(condition, step.ExecutionContext.ExpressionValues, step.ExecutionContext.ExpressionFunctions, step.ExecutionContext.ToExpressionState());
+                    }
+                    catch (Exception ex)
+                    {
+                        Trace.Info("Caught exception from expression.");
+                        Trace.Error(ex);
+                        conditionEvaluateError = ex;
+                    }
+                }
+                if (!conditionResult && conditionEvaluateError == null)
+                {
+                    // Condition is false
+                    Trace.Info("Skipping step due to condition evaluation.");
+                    step.ExecutionContext.Result = TaskResult.Skipped;
+                    continue;
+                }
+                else if (conditionEvaluateError != null)
+                {
+                    // Condition error
+                    step.ExecutionContext.Error(conditionEvaluateError);
+                    step.ExecutionContext.Result = TaskResult.Failed;
+                    ExecutionContext.Result = TaskResult.Failed;
+                    break;
+                }
+                else
+                {
+                    await RunStepAsync(step);
+                }
 
                 // Check failed or canceled
                 if (step.ExecutionContext.Result == TaskResult.Failed || step.ExecutionContext.Result == TaskResult.Canceled)
                 {
+                    Trace.Info($"Update job result with current composite step result '{step.ExecutionContext.Result}'.");
                     ExecutionContext.Result = step.ExecutionContext.Result;
-                    break;
+                    ExecutionContext.Root.Result = TaskResultUtil.MergeTaskResults(ExecutionContext.Root.Result, step.ExecutionContext.Result.Value);
```

Comment:
Notice we removed the break. In the past we would short circuit running composite steps here. Now we update the JobContext, so when we evaluate the condition for the next step (defaults to `success()` it will fail).

We have to do this because we may have multiple post steps with unique conditions. If the first post step fails, and the second has a condition of `failure()` we need to ensure we actually run the second. So the job status needs to be up to date.

The goal here is to mirror how these steps would have ran if they were inside a workflow file and not a composite action.